text/0002-binary-distribution-format.md

```diff
@@ -0,0 +1,367 @@
+    extension programmers and new extension programmers.
+*   Discuss how this impacts the ability to develop, build and distribute
+    extension packages.
+ -->
+
+## File Format
+
+### File name convention
+
+The trunk filename is:
+
+```
+{package}-{version}+{pg}-{platform}.trunk
+```
+
+Definition of variables:
+
+*   `package`: Package name, e.g. `pgmq`, `postgis`, `pgAdmin`, `pg_top`.
+*   `version`: Distribution version in [SemVer] format without build metadata,
+    e.g., `0.8.6` or `1.0.0-beta`.
+*   `pg`: Major version of Postgres the binary was built against, e.g.,
+    `pg15`, `pg16`.
+*   `platform`: The platform the binary was built for. Will be made up of one
+    to three hyphen-delimited[^hyphen] values for the OS, version
+    information[^PEPs], and CPU architecture. Examples: `any`,
+    `gnulinux-amd64`, `darwin-23.5.0.arm64`, `musllinux-1.2-amd64v3`. The
+    allowed values will be defined one or more separate RFCs.
+
+#### Examples:
+
+*   `pgtap-1.0.1+pg15-any.trunk` packages `pgtap` version 1.9.1, compatible
+    with Postgres 15 (any minor release) on any platform.
+*   `pair-0.32.1+pg16-gnulinux-amd64.trunk` packages `pair` version 0.32.1,
+    compatible with Postgres 16 (any minor release) on GNU libc-based Linux
+    for amd64 CPUs.
+*   `pair-0.32.1+pg16-darwin-23.5.0-arm64.trunk` packages `pair` version
+    0.32.1, compatible with Postgres 16 (any minor release) on Darwin version
+    23.5.0 (macOS) for arm64 CPUs.
+
+#### Escaping and Unicode
+
+The `+` in the file name indicates the division between the package name and
+version and the package metadata. The package name and version must not
+include a `+`. This allows the file name, without the `.trunk` extension, to
+also function as a valid [SemVer].
+
+Tools producing trunks should verify that the filename components do not
+contain `+`, as the resulting file may not be processed correctly if it does.
+
+The package name should be lowercase and the file name components should all
+be UTF-8.
+
+The filenames *inside* the archive are encoded as UTF-8. Although some ZIP
+clients in common use do not properly display UTF-8 filenames, the encoding is
+supported by the ZIP specification.
+
+#### Parsing
+
+Parsing of the file name takes place in four parts:
+
+1.  For the file name, remove the `.trunk` extension. If working with the
+    directory name (prefix) extracted from the archive, there will be no
+    `.trunk` extension.
+
+2.  Split the name into two parts at the `+` sign. The left part is the
+    package name and [SemVer]. The right part is the platform specification.
+
+3.  For the left part, split on the right-most dash. If the string to the
+    right of the dash is a valid [SemVer], then the left part is the package
+    name. If the right string is not a valid [SemVer], try again at the second
+    right-most dash and check again. Continue until a valid SemVer is produced
+    or else fail.
```

Comment:
Semver are required to have 3 parts, with any suffix after a dash (presumably not with another semver-like system, so `3.1.2-0` would be valid, but `3.1.2-1.2.3`?  This part feels like it has the potential for some ambiguity to me to be honest, but I trust you're more familiar with semver than I am at this point. :-)

Reply:
Bah, no, of course, `3.1.2-1.2.3` is a valid Semver.

So perhaps it instead needs to be that distribution names cannot start with a digit or use a digit after a dash. That would simplify things: the version starts at the first digit after a dash. Of existing distributions, it looks like only `DBT-2` would be a problem:

```psql
distinct name from distributions where name LIKE '%-%';
           name           
──────────────────────────
 DBT-2
 E-Maj
 foreign-keycloak-wrapper
 hashtypes-aleksabl
 neo4j-fdw
 passwd-fdw
 pg-json
 Slony-I
 soundex-function
 soundex-operator
 soundex-type
 trunklet-format
(12 rows)
```

Note that this does not apply to the name of the extensions, apps, or anything else inside the distribution, just the name of the thing being distributed.

Reply:
Okay, added the [Package Name](/pgxn/rfcs/pull/3/files#diff-3a6d7e2784cf9bc361c4255353765fc71bc523285341b9f77b6787223165a19cR138-R145) type to #3. I think that will solve the problem. I'll rewrite this bit.